### PR TITLE
New Feature: fft_aggregated

### PR DESCRIPTION
**Pull request by aloosley**
Hi all,

As mentioned in the [Spectral Centroid PR](https://github.com/blue-yonder/tsfresh/pull/314), I decided it would be cleaner to just restart the PR.  As requested, I have added a new combiner feature called fft_aggregated.  

I added tests for it, they pass.  There are other tests that I'm not sure are related to this new feature that are not passing for me.

Best.

### CODE REVIEW

File: tsfresh/feature_extraction/feature_calculators.py [hunk 1]
```diff
@@ -826,6 +826,53 @@ def complex_agg(x, agg):
...
+    :type param: list
+    :return: the different feature values
+    :return type: pandas.Series
+    """
+
+    assert set([config["aggtype"] for config in param]) <= set(["centroid", "variance", "skew", "kurtosis"]), \
+        'Attribute must be "centroid", "variance", "skew", "kurtosis"'
+
+
+    def get_moment(y, moment):
```
Comment (MaxBenChrist) [inline tsfresh/feature_extraction/feature_calculators.py]: can you please add a docstring here?

File: tsfresh/feature_extraction/feature_calculators.py [hunk 2]
```diff
@@ -826,6 +826,53 @@ def complex_agg(x, agg):
...
+        return y.dot(np.arange(len(y)) ** moment) / y.sum()
+
+    def get_centroid(y):
+        return get_moment(y, 1)
+
+    def get_variance(y):
```
Comment (MaxBenChrist) [inline tsfresh/feature_extraction/feature_calculators.py]: can you please add a docstring here?

File: tsfresh/feature_extraction/feature_calculators.py [hunk 3]
```diff
@@ -826,6 +826,53 @@ def complex_agg(x, agg):
...
+    """
+
+    assert set([config["aggtype"] for config in param]) <= set(["centroid", "variance", "skew", "kurtosis"]), \
+        'Attribute must be "centroid", "variance", "skew", "kurtosis"'
+
+
+    def get_moment(y, moment):
+        return y.dot(np.arange(len(y)) ** moment) / y.sum()
+
+    def get_centroid(y):
```
Comment (MaxBenChrist) [inline tsfresh/feature_extraction/feature_calculators.py]: can you please add a docstring here?

File: tsfresh/feature_extraction/feature_calculators.py [hunk 4]
```diff
@@ -826,6 +826,53 @@ def complex_agg(x, agg):
...
+    def get_variance(y):
+        return get_moment(y, 2) - get_centroid(y) ** 2
+
+    def get_skew(y):
+        return skewness(y)
+
+    def get_kurtosis(y):
+        return kurtosis(y)
+
+    calculation = dict(
```
Comment (MaxBenChrist) [inline tsfresh/feature_extraction/feature_calculators.py]: I like the way how you calculate the attributes by help of the calculation dictionary

File: tests/units/feature_extraction/test_feature_calculations.py [hunk 5]
```diff
@@ -473,6 +473,47 @@ def test_fft_coefficient(self):
...
+            {"aggtype": "variance"},
+            {"aggtype": "skew"},
+            {"aggtype": "kurtosis"}
+        ]
+        expected_index = ['aggtype_centroid', 'aggtype_variance', 'aggtype_skew', 'aggtype_kurtosis']
+
+        x = range(10)
+        res = pd.Series(dict(fft_aggregated(x, param)))
+        six.assertCountEqual(self, list(res.index), expected_index)
+        self.assertAlmostEqual(res['aggtype_centroid'], 1.135, places=3)
```
Comment (MaxBenChrist) [inline tests/units/feature_extraction/test_feature_calculations.py]: This looks like you ran the method on the input and fixated the result (in this case `1.135`).

So, this unit test does not check if the calculation is right. Instead it checks if the value of the feature is not changing between tsfresh version.

I would prefer a result that is calculated by hand. Can you think of a border case, where we can calculate the feature by hand?

File: tsfresh/feature_extraction/feature_calculators.py [hunk 6]
```diff
@@ -826,6 +826,53 @@ def complex_agg(x, agg):
...
+        centroid=get_centroid,
+        variance=get_variance,
+        skew=get_skew,
+        kurtosis=get_kurtosis
+    )
+
+    fft_abs = abs(np.fft.rfft(x))
+
+    res = [calculation[config["aggtype"]](fft_abs) for config in param]
+    index = ['aggtype_{}'.format(config["aggtype"]) for config in param]
```
Comment (MaxBenChrist) [inline tsfresh/feature_extraction/feature_calculators.py]: the error is related to the inferring of a feature calculators parameter from the column name.

To correct this, you need to add quotation marks around the string paramter, it should be 

```
index = ['aggtype_"{}"'.format(config["aggtype"]) for config in param]
```

**Comment by MaxBenChrist** [comment]
Great work @aloosley, we are close to merge this into master 👍

File: tsfresh/feature_extraction/feature_calculators.py (hunk 1 above)
Comment (aloosley) [reply tsfresh/feature_extraction/feature_calculators.py]: Done

File: tsfresh/feature_extraction/feature_calculators.py (hunk 3 above)
Comment (aloosley) [reply tsfresh/feature_extraction/feature_calculators.py]: Done

File: tsfresh/feature_extraction/feature_calculators.py (hunk 2 above)
Comment (aloosley) [reply tsfresh/feature_extraction/feature_calculators.py]: Done

File: tsfresh/feature_extraction/feature_calculators.py (hunk 4 above)
Comment (aloosley) [reply tsfresh/feature_extraction/feature_calculators.py]: Thanks.

File: tsfresh/feature_extraction/feature_calculators.py (hunk 6 above)
Comment (aloosley) [reply tsfresh/feature_extraction/feature_calculators.py]: Done.  I updated the tests accordingly.

Comment (MaxBenChrist) [reply tsfresh/feature_extraction/feature_calculators.py]: this line was not touched, was it?

File: tests/units/feature_extraction/test_feature_calculations.py [hunk 7]
```diff
@@ -473,6 +473,47 @@ def test_fft_coefficient(self):
...
         self.assertIsNaN(res['coeff_10__attr_"real"'])
 
+    def test_fft_aggregated(self):
+        param = [
+            {"aggtype": "centroid"},
+            {"aggtype": "variance"},
+            {"aggtype": "skew"},
+            {"aggtype": "kurtosis"}
+        ]
+        expected_index = ['aggtype_centroid', 'aggtype_variance', 'aggtype_skew', 'aggtype_kurtosis']
```
Comment (MaxBenChrist) [inline tests/units/feature_extraction/test_feature_calculations.py]: the expected index needs to have quotation marks around the attributes

Comment (aloosley) [reply tests/units/feature_extraction/test_feature_calculations.py]: Ok, I will push now, please note the tests will fail because I'm in the middle of fixing the kurtosis and skew calculations (it was incorrect to use the pandas.Series skew/kurtosis functions that I had committed).

**Comment by aloosley** [comment]
Please give some time for me to finish fixing the skew and kurtosis calculations and to add a few tests.

**Comment by MaxBenChrist** [comment]
> Please give some time for me to finish fixing the skew and kurtosis calculations and to add a few tests.

Sure sure, take your time :)

**Comment by aloosley** [comment]
Everything should be set now.  I also fixed kurtosis and skew calculations for the edge cases where the discrete calculations blow up and are no longer meaningful. For example, with a dirac delta, the 0/0 division for kurtosis should limit to 3 but gives nonsense in the real world case because of noise and rounding. I also added hand calculated test values for x=sin() and x=Normal() cases.

**Comment by aloosley** [comment]
One final push to I just fixed up a few bad comments I had glossed over.

**Comment by MaxBenChrist** [comment]
Wow. Great work @aloosley! 

Happy to merge this one!

**Comment by aloosley** [comment]
Thanks for the back and forth with me.  That was great.

**Comment by MaxBenChrist** [comment]
> Thanks for the back and forth with me. That was great.

Thanks. Looking forward to see more feature calculators from you ;)
